File: 108_MaximumSubarray.cpp
// Topic: Maximum Subarray
// Description: Find contiguous subarray with largest sum.
// Approach: Kadane's algorithm (DP in O(n) time, O(1) space).

#include <bits/stdc++.h>
using namespace std;

int maxSubArray(vector<int>& nums)
{
    int best = INT_MIN, cur = 0;
    for (int x : nums)
    {
        cur = max(x, cur + x);
        best = max(best, cur);
    }
    return best;
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    vector<vector<int>> tests = {{-2, 1, -3, 4, -1, 2, 1, -5, 4}, {1}, {-1, -2, -3}};

    for (int i = 0; i < (int)tests.size(); ++i)
    {
        cout << "Test " << (i + 1) << ": " << maxSubArray(tests[i]) << '\n';
    }
    return 0;
}

/*
Time Complexity: O(n)
Space Complexity: O(1)
*/

File: 108_MaximumSubarray.py
# Topic: Maximum Subarray
# Description: Find contiguous subarray with largest sum.
# Approach: Kadane's algorithm (DP in O(n) time, O(1) space).

from typing import List


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        best = float('-inf')
        cur = 0
        for x in nums:
            cur = max(x, cur + x)
            best = max(best, cur)
        return best


if __name__ == "__main__":
    tests = [[-2, 1, -3, 4, -1, 2, 1, -5, 4], [1], [-1, -2, -3]]
    for i, nums in enumerate(tests):
        print(f"Test {i + 1}: {Solution().maxSubArray(nums)}")

"""
Time Complexity: O(n)
Space Complexity: O(1)
"""
